### PR TITLE
kernel/sched: Make sure the lowest priority is in the range of 0 to 127

### PR DESCRIPTION
**Pull request by yashi**
When preemption is enabled, we define K_LOWEST_THREAD_PRIO as
CONFIG_NUM_PREEMPT_PRIORITIES, which is limited from 0 to 128 by
kernel/Kconfig.  Unfortunately, using the number of priorities as the
thread priority is wrong.  The range of the priority for preemptible
threads must be 0 to 127.

K_LOWEST_THREAD_PRIO == 128 is currently used to create the idle
thread in init_idle_thread() when CONFIG_NUM_PREEMPT_PRIORITIES is
128.  The number 128 is passed down to z_setup_new_thread() ->
z_new_thread() -> z_new_thread_init() -> z_init_thread_base(), and
finally assigned to thread_base->prio which is type of s8_t.

Assigning 128 of type int to s8_t results in -127, which in turn means
now that the idle thread is the highest priority thread of all.  We
didn't notice this because we had special check for the idle thread in
should_preempt(), which is removed by the commit bd049626c5b04.

This patch only changes the definition of K_LOWEST_THREAD_PRIO but it
might be better to have some range check before the assignment to
thread_base->prio in z_init_thread_base() and z_thread_priority_set().

Signed-off-by: Yasushi SHOJI <y-shoji@ispace-inc.com>

### CODE REVIEW

**Review by tyrantlizard77 — APPROVED** [review]

**Comment by yashi** [comment]
@andrewboie Thanks for the review.  I noticed that the `sanitycheck` failed for `samples/philosophers/sample.philosopher.preempt_only`.

This is because it loops like:
```C
	for (int i = 0; i < NUM_PHIL; i++) {
		int prio = new_prio(i);

		k_thread_create(&threads[i], &stacks[i][0], STACK_SIZE,
				philosopher, (void *)i, NULL, NULL, prio,
				K_USER, K_FOREVER);
```
where `#define NUM_PHIL 6` and `CONFIG_NUM_PREEMPT_PRIORITIES=6`.  Because the system only has total of 6 preemptible threads but Zephyr reserve one for system (the idle thread), application can only use 5 priorities, as `#define K_LOWEST_APPLICATION_THREAD_PRIO (K_LOWEST_THREAD_PRIO - 1).

So when priority `5` is given to `k_thread_create()`, `Z_ASSERT_VALID_PRIO()` macro in `z_new_thread()` fails.

Am I understanding OK?  If so, should I add another commit to the PR or should I squash it to the current commit?

That change should be like this:
```
diff --git a/samples/philosophers/sample.yaml b/samples/philosophers/sample.yaml
index 79e501ad97..d8b6919468 100644
--- a/samples/philosophers/sample.yaml
+++ b/samples/philosophers/sample.yaml
@@ -33,7 +33,7 @@ tests:
     extra_args: "-DFORKS=LIFOS"
   sample.philosopher.preempt_only:
     extra_configs:
-      - CONFIG_NUM_PREEMPT_PRIORITIES=6
+      - CONFIG_NUM_PREEMPT_PRIORITIES=7
       - CONFIG_NUM_COOP_PRIORITIES=0
       - CONFIG_BT=n
   sample.philosopher.coop_only:
```

**Comment by yashi** [comment]
https://docs.zephyrproject.org/latest/reference/kernel/threads/index.html?highlight=priority#thread-priorities
> The kernel supports a virtually unlimited number of thread priority levels.
Should we address this as well?  We only support 256 thread priorities.  Yes, for many applications, 256 is a big number but this line is a left over from the era we had 32 bit for priority, which is changed by the commit f9554765596b7.  The quoted line is predates the change and introduced by the commit e135a273ecb34.

**Review by andyross — APPROVED** [review]
This all looks correct.  Those priority macros are sort of a confusing mess, honestly, but I didn't want to touch them during the scheduler work because they're in kernel.h and technically part of our supported API.

There's no particular reason for the field size in struct k_thread to be the limit here, though.  A very reasonable feature addition would just be to stretch this (though note that we need some  high bits reserved for the quirky sched_lock recursion count trick we play).  Only SCHED_MULTIQ has any per-priority-level overhead, and it has a firm limit of 32 priorities right now (and enforces it via a build assert).

File: doc/reference/kernel/threads/index.rst [hunk 1]
```diff
@@ -348,18 +348,33 @@ A thread's initial priority value can be altered up or down after the thread
 has been started. Thus it possible for a preemptible thread to become
 a cooperative thread, and vice versa, by changing its priority.
 
-The kernel supports a virtually unlimited number of thread priority levels.
+The kernel supports, at the maximum, a total of 256 thread priority levels.
```
Comment (dbkinder) [inline doc/reference/kernel/threads/index.rst]: simpler would be:
```
The kernel supports up to 256 thread priority levels.
```

File: doc/reference/kernel/threads/index.rst [hunk 2]
```diff
@@ -348,18 +348,33 @@ A thread's initial priority value can be altered up or down after the thread
...
 The configuration options :option:`CONFIG_NUM_COOP_PRIORITIES` and
 :option:`CONFIG_NUM_PREEMPT_PRIORITIES` specify the number of priority
-levels for each class of thread, resulting the following usable priority
-ranges:
+levels for each class of thread in the system, resulting the following
```
Comment (dbkinder) [inline doc/reference/kernel/threads/index.rst]:
```suggestion
levels for each class of thread in the system, yielding the following
```

**Review by dbkinder — DISMISSED** [review]

File: doc/reference/kernel/threads/index.rst (hunk 1 above)
Comment (yashi) [reply doc/reference/kernel/threads/index.rst]: Done. Thanks.

File: doc/reference/kernel/threads/index.rst (hunk 2 above)
Comment (yashi) [reply doc/reference/kernel/threads/index.rst]: Thanks. Updated.

**Comment by yashi** [comment]
Could someone please help me resolve the error from `test_posix_pthread_termination()`?  Or, tell me how to run the test on my env?  When I `make run` for `tests/posix/common/`, I got
```
SeaBIOS (version rel-1.12.0-0-ga698c8995f-prebuilt.qemu.org)
Booting from ROM..
```
and stops.

`./scripts/sanitycheck --ninja -s tests/posix/common/portability.posix` also fails...

**Comment by yashi** [comment]
> Those priority macros are sort of a confusing mess, honestly, but I didn't want to touch them during the scheduler work because they're in kernel.h and technically part of our supported API.

I can now see what you mean.  Or I just don't know the spec of Zephyr's thread priority.

I mean, do we _always_ reserve the lowest thread priority for the idle thread?  What if we have only one preempt priority or one coop priority?  Is such a configuration invalid?

If we must reserve the lowest thread priority for the idle thread, the following could be a rule

- BOTH: num coop must be 1 or more, num preempt must be 2 or more
- COOP only: num coop must be 2 or more, num preempt must be 0
- PREEMPT only: num coop must be 0, num preempt must be 2 or more

IIOW, the rule is " for applications, we reserve at least one priority in the given class".
With this rule, the priorities for kernel threads, applications threads, and the idle thread are

- BOTH: kernel and application threads must be anything except the lowest, the idle must be the lowest (num preempt - 1).  For applications, the number of coop priorities is `CONFIG_NUM_COOP_PRIORITIES` and the number of preempt priorities is `CONFIG_NUM_PREEMPT_PRIORITIES - 1`.
- COOP only: kernel and application can be anything except the lowest, idle must be the lowest (-1). the number of priorities for applications is `CONFIG_NUM_COOP_PRIORITIES - 1`.
- PREEMPT only: kernel and application can be anything except the lowest, idle must be num preempt.  the number of priorities for applications is `CONFIG_NUM_PREEMPT_PRIORITIES - 1`.

**Comment by carlescufi** [comment]
@yashi try using `-DBOARD=mps2_an385` assuming you are on Linux and with the latest SDK

**Review by ioannisg — APPROVED** [review]

**Comment by andyross** [comment]
> I mean, do we always reserve the lowest thread priority for the idle thread? What if we have only one preempt priority or one coop priority? Is such a configuration invalid?

It's actually not required in the current scheduler that the idle thread get its own priority, though it's obviously untested.  The idle threads get chosen if there's no thread available in the queue, they aren't strictly involved in priority comparison logic.  The macros still enforce it because of history.

**Comment by ioannisg** [comment]
@yashi is this PR still relevant?

**Comment by yashi** [comment]
It still is an issue.  A work around was simple enough (to just change the number of preemptive threads), but a proper fix is way too complicated (to define how we want our scheduler to be).

So one way to do is to document as a known bug.

I'll give another look and ask if I need an advice.

**Comment by ioannisg** [comment]
Moving to 2.1 since it is not ready to go into 2.0.
@yashi would you be able to resolve conflicts?

**Comment by yashi** [comment]
I'll take a look at it.

**Review by dbkinder — APPROVED** [review]
+1 for doc changes

**Comment by galak** [comment]
@yashi any plans to look at this?

**Comment by yashi** [comment]
I'd love to work on it but I'm still working on other parts of the kernel right now.  And the workaround is good enough for me not to fix it yet.  We can close this for now and I'll re-open it when I get back to it.

**Pull request closed without merging**